File: src/core/context.ts
import { v4 as uuidv4 } from "uuid";
import {
  ContextEnvelope, BranchType, ContextStatus, Message,
  MMCP_VERSION, MergeStrategy
} from "./types";

export function createContext(params: {
  task: string;
  role: string;
  model: string;
  parent_ids?: string[];
  branch_type?: BranchType;
  history?: Message[];
  system_prompt?: string;
  depth?: number;
  shard_index?: number;
  merge_strategy?: MergeStrategy;
  max_retries?: number;
  metadata?: Record<string, unknown>;
  required_skills?: string[];
  matched_skills?: string[];
  missing_skills?: string[];
}): ContextEnvelope {
  return {
    mmcp_version: MMCP_VERSION,
    id: `ctx_${uuidv4().replace(/-/g, "").slice(0, 10)}`,
    parent_ids: params.parent_ids ?? [],
    children: [],
    task: params.task,
    history: params.history ?? [],
    system_prompt: params.system_prompt,
    model: params.model,
    role: params.role,
    branch_type: params.branch_type ?? "handoff",
    depth: params.depth ?? 0,
    shard_index: params.shard_index,
    merge_strategy: params.merge_strategy,
    status: "pending",
    retry_count: 0,
    max_retries: params.max_retries ?? 2,
    tokens_used: undefined,
    created_at: new Date().toISOString(),
    metadata: params.metadata ?? {},
    required_skills: params.required_skills,
    matched_skills: params.matched_skills,
    missing_skills: params.missing_skills,
  };
}

// Build the history for a context by collecting parent outputs
export function buildHistory(
  parentContexts: ContextEnvelope[],
  task: string,
  role: string
): Message[] {
  const history: Message[] = [];

  if (parentContexts.length === 1) {
    // Simple handoff: inherit parent history + append output
    const parent = parentContexts[0];
    history.push(...parent.history);
    if (parent.output) {
      history.push({ role: "assistant", content: parent.output });
    }
  } else if (parentContexts.length > 1) {
    // Merge: collect all parent outputs as a structured user message
    const parts = parentContexts
      .filter(p => p.output)
      .map(p => `[${p.role.toUpperCase()}]:\n${p.output}`)
      .join("\n\n---\n\n");

    history.push({
      role: "user",
      content: `You are the ${role}. Multiple agents have completed their tasks. Here are their outputs:\n\n${parts}\n\nTask: ${task}`
    });
  }

  return history;
}

// Topological sort of context DAG
export function topologicalSort(contexts: ContextEnvelope[]): ContextEnvelope[] {
  const map = new Map(contexts.map(c => [c.id, c]));
  const visited = new Set<string>();
  const result: ContextEnvelope[] = [];

  function visit(id: string) {
    if (visited.has(id)) return;
    const ctx = map.get(id);
    if (!ctx) return;
    // Visit parents first
    for (const pid of ctx.parent_ids) {
      visit(pid);
    }
    visited.add(id);
    result.push(ctx);
  }

  for (const ctx of contexts) {
    visit(ctx.id);
  }

  return result;
}

// Check if all parents of a context are done
export function parentsReady(
  context: ContextEnvelope,
  allContexts: Map<string, ContextEnvelope>
): boolean {
  if (context.parent_ids.length === 0) return true;
  return context.parent_ids.every(pid => {
    const parent = allContexts.get(pid);
    return parent?.status === "done";
  });
}

File: python/mmcp_core/context.py
"""
MMCP Context — mirrors TypeScript src/core/context.ts exactly.
Context creation, history building, topological sort with cycle detection.
"""
from __future__ import annotations
import uuid
from datetime import datetime, timezone
from .types import (
    ContextEnvelope, BranchType, Message,
    MergeStrategy, MMCP_VERSION
)


def create_context(
    task: str,
    role: str,
    model: str,
    parent_ids: list[str] | None = None,
    branch_type: BranchType = "handoff",
    history: list[Message] | None = None,
    system_prompt: str | None = None,
    depth: int = 0,
    shard_index: int | None = None,
    merge_strategy: MergeStrategy | None = None,
    max_retries: int = 2,
    metadata: dict | None = None,
    required_skills: list[str] | None = None,
    matched_skills: list[str] | None = None,
    missing_skills: list[str] | None = None,
) -> ContextEnvelope:
    """Create a new ContextEnvelope with full 128-bit UUID."""
    ctx_id = f"ctx_{uuid.uuid4().hex}"
    return ContextEnvelope(
        mmcp_version=MMCP_VERSION,
        id=ctx_id,
        parent_ids=parent_ids or [],
        children=[],
        task=task,
        history=history or [],
        model=model,
        role=role,
        branch_type=branch_type,
        depth=depth,
        shard_index=shard_index,
        merge_strategy=merge_strategy,
        status="pending",
        retry_count=0,
        max_retries=max_retries,
        created_at=datetime.now(timezone.utc).isoformat(),
        metadata=metadata or {},
        system_prompt=system_prompt,
        required_skills=required_skills,
        matched_skills=matched_skills,
        missing_skills=missing_skills,
    )


def build_history(
    parent_contexts: list[ContextEnvelope],
    task: str,
    role: str,
) -> list[Message]:
    """Build history by collecting parent outputs."""
    history: list[Message] = []

    if len(parent_contexts) == 1:
        parent = parent_contexts[0]
        history.extend(parent.history)
        if parent.output:
            history.append(Message(role="assistant", content=parent.output))
    elif len(parent_contexts) > 1:
        parts = "\n\n---\n\n".join(
            f"[{p.role.upper()}]:\n{p.output}"
            for p in parent_contexts if p.output
        )
        history.append(Message(
            role="user",
            content=(
                f"You are the {role}. Multiple agents completed their tasks. "
                f"Here are their outputs:\n\n{parts}\n\nTask: {task}"
            ),
        ))

    return history


def topological_sort(contexts: list[ContextEnvelope]) -> list[ContextEnvelope]:
    """Topological sort with cycle detection via visiting set."""
    ctx_map = {c.id: c for c in contexts}
    visited: set[str] = set()
    visiting: set[str] = set()  # tracks current DFS path
    result: list[ContextEnvelope] = []

    def visit(ctx_id: str) -> None:
        if ctx_id in visited:
            return
        if ctx_id in visiting:
            raise ValueError(
                f"MMCP cycle detected: context {ctx_id} is its own ancestor. "
                f"DAG must be acyclic. Check parent_ids for circular references."
            )
        ctx = ctx_map.get(ctx_id)
        if not ctx:
            return

        visiting.add(ctx_id)
        for pid in ctx.parent_ids:
            visit(pid)
        visiting.remove(ctx_id)
        visited.add(ctx_id)
        result.append(ctx)

    for ctx in contexts:
        visit(ctx.id)

    return result


def parents_ready(
    context: ContextEnvelope,
    all_contexts: dict[str, ContextEnvelope],
) -> bool:
    """Check if all parents of a context are done."""
    if not context.parent_ids:
        return True
    return all(
        all_contexts.get(pid) is not None and all_contexts[pid].status == "done"
        for pid in context.parent_ids
    )
